ibis/sql/postgres/udf/api.py

```diff
@@ -0,0 +1,218 @@
+import inspect
+from decimal import Decimal
+import collections
+import itertools
+
+import sqlalchemy
+
+import ibis.expr.rules as rlz
+from ibis.expr import datatypes
+from ibis.expr.signature import Argument as Arg
+from ibis.sql.postgres.compiler import PostgresUDFNode, add_operation
+
+
+_udf_name_cache = collections.defaultdict(itertools.count)
+
+
+# type mapping based on: https://www.postgresql.org/docs/10/plpython-data.html
+sql_default_type = 'VARCHAR'
+
+pytype_sql = {
+    bool: 'BOOLEAN',
+    int: "INTEGER",
+    float: 'DOUBLE',
+    Decimal: 'NUMERIC',
+    bytes: 'BYTEA',
+    str: sql_default_type,
+}
+
+pytype_ibistype = {
+    bool: datatypes.Boolean(),
+    int: datatypes.Int32(),
+    float: datatypes.Float(),
+    Decimal: datatypes.Float(),
+    bytes: datatypes.Binary(),
+    str: datatypes.String(),
+}
+
+ibistype_pytype = {v: k for k, v in pytype_ibistype.items()}
+
+
+def get_sqltype(type_):
+    """Map input to the string specifying the Postgres data type
+    (as is used in SQL defining UDF signatures)
+
+    :param type_: str, a Python data type,
+                  or an ibis DataType (instance or class)
```

Comment:
This is handling too many input types. This function should accept an ibis datatype, and any validation should be done in the caller.

Reply:
Yeah, I agree that accepting too many input types is not great. I was trying to mimic the multiple dispatch in `ibis.bigquery.datatypes` without resorting to the same `multipledispatch` approach used there. I'll simplify this to just take ibis datatypes.

Reply:
Multiple dispatch no longer applicable, because I'm now only allowing ibis DataTypes as parameters, and I changed the way I'm translating to the SQL type strings.